crates/api/src/lib.rs

```diff
@@ -141,6 +150,52 @@ pub(crate) fn build_totp_2fa(
   .with_lemmy_type(LemmyErrorType::CouldntGenerateTotp)
 }
 
+/// Since removals and bans are only federated for local users,
+/// you also need to send bans for their content to local communities.
+/// See https://github.com/LemmyNet/lemmy/issues/4118
+#[tracing::instrument(skip_all)]
+pub(crate) async fn send_bans_and_removals_to_local_communities(
+  local_user_view: &LocalUserView,
+  target: &Person,
+  reason: &Option<String>,
+  remove_data: &Option<bool>,
+  context: &Data<LemmyContext>,
+) -> LemmyResult<()> {
+  let posts_community_ids =
+    Post::list_creators_local_community_ids(&mut context.pool(), target.id).await?;
+  let comments_community_ids =
+    Comment::list_creators_local_community_ids(&mut context.pool(), target.id).await?;
+
+  let ids = [posts_community_ids, comments_community_ids]
+    .concat()
+    .into_iter()
+    .unique();
+
+  for community_id in ids {
+    let ban_from_community = BanFromCommunity {
+      community_id,
+      person_id: target.id,
+      ban: true,
+      remove_data: *remove_data,
+      reason: reason.clone(),
+      expires: None,
+    };
+
+    ActivityChannel::submit_activity(
+      SendActivityData::BanFromCommunity {
+        moderator: local_user_view.person.clone(),
+        community_id,
+        target: target.clone(),
+        data: ban_from_community,
+      },
+      context,
+    )
+    .await?;
```

Comment:
This means that locally the user is banned from the site but not from these communities, while on other instances he is banned from communities. So there is no clean way to revert the ban, other than doing a ban and then unban in each community. Not a big problem but something to think about.

Also there should be a check so that this function only runs for remote users.

Reply:
How about instead of this community banning solution, I just send removals for all their posts and comments? Then we wouldn't have to deal with any of this banning / unbanning mess.

Reply:
We might have to accept that as a limitation. They might be able to comment /post on *their own site's version* of that community, but since they're site-banned from ours, their content won't get federated anywhere.

Reply:
Sending a removal for each post and comment could result in a lot of activities sent. So doing community bans seems preferable. Anyway its just a minor problem which can be solved later.

Reply:
I added the local user check, and also wrote these community bans locally.